### PR TITLE
Add TLE9104 output diagnostics sensor

### PR DESCRIPTION
**Pull request by benediktibk**
Implement the output diagnostics of the low side switch TLE9104 as a sensor.

### CODE REVIEW

**Review by henrikbrixandersen — COMMENTED** [review]
I think this ought to be done using an MFD. The approach taken here is the same as before we had `drivers/mfd/`.

**Comment by benediktibk** [comment]
> I think this ought to be done using an MFD. The approach taken here is the same as before we had `drivers/mfd/`.

Okay, I'm gonna focus on this PR with the changes and continue working on #67030 as soon as we have settled the basics in this one.

File: drivers/mfd/mfd_tle9104.c [hunk 1]
```diff
@@ -0,0 +1,682 @@
...
+	case 500:
+		tle9104_set_register_bits(&register_cfg, TLE9104_OFFDIAGCFG_DIAGFILTCFG_POS,
+					  TLE9104_OFFDIAGCFG_DIAGFILTCFG_LENGTH, 1);
+		break;
+	case 200:
+		tle9104_set_register_bits(&register_cfg, TLE9104_OFFDIAGCFG_DIAGFILTCFG_POS,
+					  TLE9104_OFFDIAGCFG_DIAGFILTCFG_LENGTH, 0);
+		break;
+	default:
+		LOG_ERR("invalid value %i for diagnostic filter time",
```
Comment (aasinclair) [inline drivers/mfd/mfd_tle9104.c:567]: These are calculated at compile time, so this error message isn't needed.
You could calculate the value at compile time using DT_ENUM_IDX, then you wouldn't need the switch case.

File: dts/bindings/mfd/infineon,tle9104.yaml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+  diagnostic-filter-time:
+    type: int
+    default: 500
+    enum:
+      - 2000
+      - 1000
+      - 500
+      - 200
+    description: "diagnostic filter time in us"
```
Comment (aasinclair) [inline dts/bindings/mfd/infineon,tle9104.yaml:48]: Default values need justification (e.g. `default matches the power-on reset value` or similar)

File: include/zephyr/drivers/sensor.h [hunk 3]
```diff
@@ -186,6 +186,10 @@ enum sensor_channel {
 	SENSOR_CHAN_GAUGE_DESIRED_VOLTAGE,
 	/** Desired charging current in mA */
 	SENSOR_CHAN_GAUGE_DESIRED_CHARGING_CURRENT,
+	/** Open load detected, boolean with one bit per output */
```
Comment (aasinclair) [inline include/zephyr/drivers/sensor.h:189]: These should be in a private header file

Comment (benediktibk) [reply include/zephyr/drivers/sensor.h:189]: I actually have another PR (https://github.com/zephyrproject-rtos/zephyr/pull/67030) lined up which uses the same sensor channels. Should it then still be private, but duplicated?

File: dts/bindings/mfd/infineon,tle9104.yaml (hunk 2 above)
Comment (benediktibk) [reply dts/bindings/mfd/infineon,tle9104.yaml]: done

File: drivers/mfd/mfd_tle9104.c (hunk 1 above)
Comment (benediktibk) [reply drivers/mfd/mfd_tle9104.c]: To be able to use DT_ENUM_IDX I would have to turn the enums into strings, if I understood it correctly. But these values actually represent numbers. Therefore it feels weird to turn them into strings.

I removed the default cases anyway, as they definitely won't be hit (unless there is a pointer going wild ...)

Comment (aasinclair) [reply drivers/mfd/mfd_tle9104.c]: DT_ENUM_IDX works with ints, e.g:
https://github.com/zephyrproject-rtos/zephyr/blob/f0b975032a68c2da38c34f01b836edb5e9e62b12/dts/bindings/pinctrl/ambiq%2Capollo4-pinctrl.yaml#L95-L106

File: include/zephyr/drivers/sensor.h (hunk 3 above)
Comment (aasinclair) [reply include/zephyr/drivers/sensor.h:189]: Yes, I think they should be private but duplicated

Comment (benediktibk) [reply include/zephyr/drivers/sensor.h]: done

File: drivers/mfd/mfd_tle9104.c (hunk 1 above)
Comment (benediktibk) [reply drivers/mfd/mfd_tle9104.c]: done

**Review by aasinclair — DISMISSED** [review]

**Review by aasinclair — DISMISSED** [review]

**Comment by benediktibk** [comment]
Rebased onto main to resolve merge conflicts.

**Comment by benediktibk** [comment]
Force pushed a change to set val2 as well to avoid random garbage.

**Comment by benediktibk** [comment]
Changed the order of initialization, it is necessary to configure the diagnostics before OUT_EN is set.

**Comment by benediktibk** [comment]
@MaureenHelm Could you please take a look at this PR?

**Comment by benediktibk** [comment]
Rebased onto main to resolve merge conflicts.

**Review by MaureenHelm — DISMISSED** [review]

**Review by aasinclair — DISMISSED** [review]

**Comment by benediktibk** [comment]
Rebased onto main to resolve merge conflicts.

**Review by aasinclair — DISMISSED** [review]

**Comment by benediktibk** [comment]
Rebased onto main to resolve merge conflicts.

**Review by MaureenHelm — APPROVED** [review]

**Review by aasinclair — APPROVED** [review]

**Pull request merged**